File: 18-Binary-Tree-Part-II/python/07_binary_tree_zigzag_level_order_traversal.py
from typing import List, Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []

        def dfs(node: Optional[TreeNode], level: int):
            if not node:
                return

            if level == len(result):
                result.append([])

            if level % 2 == 0:
                result[level].append(node.val)
            else:
                result[level].insert(0, node.val)

            dfs(node.left, level + 1)
            dfs(node.right, level + 1)

        dfs(root, 0)
        return result

File: 18-Binary-Tree-Part-II/cpp/07-binary-tree-zigzag-level-order-traversal.cpp
#include <vector>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void dfs(TreeNode* node, int level, vector<vector<int>>& result) {
        if (!node) return;

        if (level == result.size()) {
            result.push_back(vector<int>());
        }

        if (level % 2 == 0) {
            result[level].push_back(node->val);
        } else {
            result[level].insert(result[level].begin(), node->val);
        }

        dfs(node->left, level + 1, result);
        dfs(node->right, level + 1, result);
    }

public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> result;
        dfs(root, 0, result);
        return result;
    }
};
